Added steps for integrating IAM roles with GCP (#238)
astronomer/docs: enterprise/next/05_customize-airflow/06_integrate-iam.md

- Create an IAM Role
- Create a Trust Relationship

### Step 1: Create an IAM OIDC identity provider

1. Retrieve your EKS cluster's OIDC issuer URL with the following AWS CLI commands:

    ```bash
    $ aws eks list-clusters
    {
        "clusters": [

>             "astronomer-cluster"

Author: Not sure, this is in our currently published doc. Potential question for @fhoda ?

>             "astronomer-cluster"

Reviewer: `astronomer-cluster` is essentially a placeholder cluster name for whatever the actual name of the cluster that  astronomer is deployed to.

>             "astronomer-cluster"

Reviewer: ```suggestion
            "<astronomer-cluster>"
```
Not sure about the style we are going for but could do soemthing like that or just clarify in the doc what `astronomer-cluster` is meant to be.

Reviewer: I could not add this suggestion to the original comment about this. https://github.com/astronomer/docs/pull/238#discussion_r572189906

        ]
    }
    ```

    ```bash
    $ aws eks describe-cluster --name astronomer-cluster --query "cluster.identity.oidc.issuer" --output text

>     https://oidc.eks.us-west-2.amazonaws.com/id/EXAMPLEA829F4B2854D8DAE63782CE90

Author: Probably, but is there a way we can tell people how to retrieve this URL? @fhoda

>     https://oidc.eks.us-west-2.amazonaws.com/id/EXAMPLEA829F4B2854D8DAE63782CE90

Reviewer: The URL on line 52 `https://oidc.eks.us-west-2.amazonaws.com/id/EXAMPLEA829F4B2854D8DAE63782CE90` will be different for the user than what we have shown here. It will essentially be the output of the command above it.

>     https://oidc.eks.us-west-2.amazonaws.com/id/EXAMPLEA829F4B2854D8DAE63782CE90

Author: Ahh got it, thought the URL was a continuation of the command itself. Thank you for clarifying.

    ```

2. Open the [IAM console](https://console.aws.amazon.com/iam/).
3. In the navigation pane, click **Identity Providers** > **Create Provider**.
4. For **Provider Type**, click **Choose a provider type** > **OpenID Connect**.
5. For **Provider URL**, use the OIDC issuer URL for your cluster.
6. For **Audience**, use `sts.amazonaws.com`.
7. Verify that the provider information is correct, and then click **Create** to create your identity provider.

For additional information, refer to [Enable IAM Roles for service accounts](https://docs.aws.amazon.com/eks/latest/userguide/enable-iam-roles-for-service-accounts.html).

### Step 2: Create an IAM policy

1. Open the IAM console at https://console.aws.amazon.com/iam/.
2. In the navigation panel, click **Policies** > **Create policy**.
3. Open the **JSON** tab.
4. In the Policy Document field, specify the permissions you'd like to apply (or restrict) to the resource in question (e.g. read / write access to an AWS S3 bucket) You can also use the visual editor to construct your own policy.

>     The following example allows permission to an S3 bucket named `astronomer-bucket`:

Author: Tried to clarify this, is it accurate @fhoda

>     The following example allows permission to an S3 bucket named `astronomer-bucket`:

Reviewer: The policy allows permission to an S3 bucket that will later be attached to the IAM Role.

- Create an IAM Role
- Create a Trust Relationship

### Step 1: Create an IAM OIDC identity provider

1. Retrieve your EKS cluster's OIDC issuer URL with the following AWS CLI commands:

    ```bash
    $ aws eks list-clusters
    {
        "clusters": [
            "<your-cluster>"
        ]

>     }

Reviewer: This part is the output of the above command `aws eks list-clusters`. Based on some of the other changes you made, thought you may also want to separate this one out?

Author: Breaking this into two separate steps, one for each command.

### Step 4: Integrate your IAM role with Astronomer

In order to apply your IAM role to any Airflow Deployment on Astronomer, you'll need to explicitly pass an annotation key to the platform. To do so:

1. Set the following in your `config.yaml` file under `astronomer.houston.config.deployments`:

    ```yaml
    serviceAccountAnnotationKey: eks.amazonaws.com/role-arn
    ```

    Example:
    ```yaml
    global:
      baseDomain: astro.mydomain.com
      tlsSecret: astronomer-tls
      postgresqlEnabled: false
    nginx:
      privateLoadBalancer: true
    astronomer:
      houston:
        config:
          deployments:
            serviceAccountAnnotationKey: eks.amazonaws.com/role-arn
    ```

2. Push the configuration change to your platform as described in [Apply a Platform Configuration Change on Astronomer](https://www.astronomer.io/docs/enterprise/stable/manage-astronomer/apply-platform-config).

### Step 5: Create or update an Airflow Deployment with an attached IAM role

1. To create a new Airflow Deployment with your IAM role attached, run the following command:

    ```sh

>     $ astro deployment create <deployment-name> --executor=celery --cloud-role=arn:aws:iam::123456789:role/example-role

Author: Yeah it's unclear when users first acquire this ID number in the steps. Maybe @fhoda could clarify

>     $ astro deployment create <deployment-name> --executor=celery --cloud-role=arn:aws:iam::123456789:role/example-role

Reviewer: That number is an AWS Account number. This entire string: `arn:aws:iam::123456789:role/example-role` can be retrieved from IAM Management Console for the Role of interest.

    ```

    Alternatively, to update an existing Airflow Deployment with your IAM role attached, run the following:

    ```sh
    $ astro deployment update <deployment-name> --cloud-role=arn:aws:iam::123456789:role/example-role
    ```

2. Confirm the role was passed successfully to Webserver, Scheduler and Worker pods within your Airflow Deployment by running the following run the following command:

    ```bash
    $ kubectl describe po <pod-name> -n <airflow-namespace>
    ```

    You should see the following in your output:

    ```yaml
    AWS_ROLE_ARN: arn:aws:iam::123456789:role/example-role
    AWS_WEB_IDENTITY_TOKEN_FILE: /var/run/secrets/eks.amazonaws.com/serviceaccount/token
    ```

> > **Note:** If using Airflow `1.10.5`, you'll need to add `boto3 >=1.9` and `botocore >= 1.12` to your `requirements.txt` file.

Author: Since this is a local Airflow file I'm guessing it doesn't require a config.yaml change

> > **Note:** If using Airflow `1.10.5`, you'll need to add `boto3 >=1.9` and `botocore >= 1.12` to your `requirements.txt` file.

Reviewer: This is just a validation, so no other changes should be needed.

    $ gcloud container clusters update <cluster-name> \
      --workload-pool=<project-id>.svc.id.goog
    ```

2. Configure your node pool to use Workload Identity by running the following command:

    ```bash
    $ gcloud container node-pools update <nodepool-name> \
      --cluster=<cluster-name> \
      --workload-metadata=GKE_METADATA
    ```

### Step 2: Create a Google service account

To create a Google service account, run the following command:

```bash
$ gcloud iam service-accounts create <gsa-name>
```

### Step 3: Configure Astronomer

Add the following to your `config.yaml` file and push it to your platform as described in [Apply a Config Change](https://www.astronomer.io/docs/enterprise/v0.23/manage-astronomer/apply-platform-config):

```yaml
astronomer:
  houston:
    config:
      deployments:
        serviceAccountAnnotationKey: iam.gke.io/gcp-service-account
```

### Step 4: Create an Airflow Deployment

1. Create an Airflow Deployment with your Google service account attached by running the following command:

    ```bash
    $ astro deployment create <deployment-name> --executor=celery --cloud-role=<gsa-name>@<project-id>.iam.gserviceaccount.com
    ```

> 2. Note the name of each GKE service account that appears when you run the following command:

Author: Added "GKE" here and in step 3, but I'm not super sure about the difference between GKE vs GCP service accounts. Could use some confirmation that this is accurate.

Reviewer: This distinction is correct. Just to note for your reference that GKE service accounts are tied to resources deployed within GKE itself and can only have permissions within GKE except when bound to a GCP service account. While GCP service accounts can have permissions across the platform but cannot define permission to resources within a GKE cluster.

> 2. Note the name of each GKE service account that appears when you run the following command:

Reviewer: ```suggestion
2. Note the name of the worker and scheduler GKE service account that appears when you run the following command:
```

    ```bash
    $ kubectl get sa -n <your-airflow-namespace>
    ```

> 3. Create an IAM policy binding your Google and GKE service accounts by running the following command for each of the GKE service accounts you noted:

Reviewer: ```suggestion
3. Create an IAM policy binding your Google and GKE service accounts by running the following command for the workers and scheduler GKE service accounts you noted:
```

    ```bash
    $ astro deployment create <deployment-name> --executor=celery --cloud-role=<gsa-name>@<project-id>.iam.gserviceaccount.com
    ```

2. Note the name of the Worker and Scheduler service accounts that appear when you run the following command:

    ```bash
    $ kubectl get sa -n <your-airflow-namespace>
    ```

3. Create an IAM policy binding your Google and GKE service accounts by running the following command for both the Worker and Scheduler GKE service accounts you noted:

    ```bash
    gcloud iam service-accounts add-iam-policy-binding \
    --role roles/iam.workloadIdentityUser \
    --member "serviceAccount:<project-id>.svc.id.goog[<your-airflow-namespace>/<gke-service-account-name>]" \
    <gsa-name>@<project-id>.iam.gserviceaccount.com
    ```

### Step 4: Confirm Workload Identity is working

1. Create an interactive session by running the following command:

    ```bash
  $ kubectl run -it \
      --image google/cloud-sdk:slim \
      --serviceaccount <worker-serviceaccount> \
      --namespace <your-airflow-namespace> \
      workload-identity-test

>     ```

Reviewer: When I preview the doc this part does not come out as a proper code block.

Author: Thank you for catching these!! They appeared rendered in Atom, but not in GitHub which is the most reliable tell.

2. In the interactive session, confirm you're able to authenticate successfully via Workload Identity by running the following command:

    ```bash
  $ gcloud auth list

>     ```

Reviewer: When I preview the doc this part does not come out as a proper code block.

}
```

    Example:

>     ```json

Reviewer: This part doesn't render the code block properly form me either. `Example:` is included in the block and the **```**  shows up around the code. Not sure how you wanted to reformat since you had indented this part.

    {
      "Version": "2012-10-17",
      "Statement": [
        {
          "Effect": "Allow",
          "Principal": {
            "Service": "ec2.amazonaws.com"
          },
          "Action": "sts:AssumeRole"
        },
        {
          "Effect": "Allow",
          "Principal": {
            "Federated": "arn:aws:iam::<your-iam-id>:oidc-provider/oidc.eks.us-west-2.amazonaws.com/id/EXAMPLEA829F4B2854D8DAE63782CE90"
          },
          "Action": "sts:AssumeRoleWithWebIdentity",
          "Condition": {
            "StringLike": {
              "oidc.eks.us-west-2.amazonaws.com/id/EXAMPLEA829F4B2854D8DAE63782CE90:sub": "system:serviceaccount:astronomer-*:*"
            }
          }
        }
      ]
    }

>     ```

Reviewer: This code block doesn't render properly for me either. I think it has something to do with the indentation of of `Example:` . Wasn't sure you wanted to adjust.
